src/policy/protect_payload.rs

```diff
@@ -138,12 +117,129 @@ impl PolicyModule for ProtectPayloadPolicy {
 }
 
 impl ProtectPayloadPolicy {
-    fn lock(&mut self, _mctx: &mut MiralisContext, _ctx: &mut VirtContext) {
-        self.protected = true;
+    fn check_interrupt(
+        &mut self,
+        ctx: &mut VirtContext,
+        mctx: &mut MiralisContext,
+    ) -> PolicyHookResult {
+        let cause = ctx.trap_info.get_cause();
+        match cause {
+            MCause::LoadAddrMisaligned | MCause::StoreAddrMisaligned => {
+                self.handle_misaligned_op(ctx, mctx)
+            }
+            MCause::EcallFromSMode => self.handle_ecall(ctx, mctx),
+            _ => PolicyHookResult::Ignore,
+        }
+    }
+
+    fn handle_misaligned_op(
+        &mut self,
+        ctx: &mut VirtContext,
+        mctx: &mut MiralisContext,
+    ) -> PolicyHookResult {
+        let instr: usize = 0;
+
+        let instr_ptr = ctx.trap_info.mepc as *const u8;
+
+        // With compressed instruction extention ("C") instructions can be misaligned.
+        // TODO: add support for 16 bits instructions
+        let instr_addr: usize = &instr as *const usize as usize;
+        self.copy_from_previous_mode(instr_ptr, instr_addr as *mut u8);
+        let instr = mctx.decode(instr);
+
+        match instr {
+            Instr::Load {
+                rd, rs1, imm, len, ..
+            } => {
+                assert!(
+                    len.to_bytes() == 8,
+                    "Implement support for other than 8 bytes misalinged accesses"
+                );
+
+                // Build the value
+                let start_addr: *const u8 =
+                    ((ctx.regs[rs1 as usize] as isize + imm) as usize) as *const u8;
+
+                let value_to_read: u64 = 0x0;
+                let addr_emulated_value = &value_to_read as *const u64 as *mut u8;
+
+                self.copy_from_previous_mode(start_addr, addr_emulated_value);
+
+                // Return the value
+                ctx.regs[rd as usize] = value_to_read as usize;
+            }
+            Instr::Store {
+                rs2, rs1, imm, len, ..
+            } => {
+                assert!(
+                    len.to_bytes() == 8,
+                    "Implement support for other than 8 bytes misalinged accesses"
+                );
+
+                // Build the value
+                let start_addr: *mut u8 =
+                    ((ctx.regs[rs1 as usize] as isize + imm) as usize) as *mut u8;
+
+                let val = ctx.regs[rs2 as usize];
+                let value_to_store = (&val) as *const usize as *mut u8;
+
+                self.copy_from_previous_mode_store(value_to_store, start_addr);
+            }
+            _ => {
+                panic!("Must be a load instruction here")
+            }
+        }
+
+        ctx.pc += 4;
+        PolicyHookResult::Overwrite
+    }
+
+    fn copy_from_previous_mode(&mut self, src: *const u8, dest: *mut u8) {
```

Comment:
You can't have this function being labelled as safe, it can cause undefined behavior depending on what pointers you pass in. So first thing mark it as unsafe.

Then avoid using pointers like this, instead prefer arrays of bytes `&[u8; 8]` that you can convert to `usize`s using `from_le_bytes` or `to_le_bytes` (see [doc](https://doc.rust-lang.org/std/primitive.u64.html#method.to_le_bytes))